docs/hugo/content/design/ADR-2022-09-Multiple-Credential-Operator.md

```diff
@@ -0,0 +1,153 @@
+
+**Fixed namespaced secret name pattern:**
+```
+Namespaced secret -> aso-credential (in the namespace)
+```
+
+**Annotation for per-resource-group and per-resource secret**
+```
+annotations: 
+  serviceoperator.azure.com/credential-from: any-secret
+```
+
+**Pros:**
+1. Users don't have to specify the configuration for each resource explicitly if using a namespaced secret
+2. Users get the flexibility of over-riding the credential they want to use as per resource and as per resource group
+
+**Cons:**
+1. Could be complicated/confusing for users to understand the usage and credentials being used
+
+### Option 4: Secret lookup + configuration using global secret
+
+This option, relates to the option 3. However, the change is how we configure to use the per-resource-group and per-resource secret. We can have a unique slice property(CREDENTIAL_FROM) in the global secret where admin users can specify the secrets they want to load on the startup.
+Then only these loaded secrets would be used further for per-resource and per-resource-group. This option would restrict the other non-admin users(who don't have access to the global secret) to add their credential. 
+
+**Pros:**
+1. Users don't have to specify the configuration for each resource explicitly if using a namespaced secret
+2. Security conscious, not anyone in the cluster can add to use their credentials.
+
+**Cons:**
+1. Could be complicated/confusing for users to understand the usage and credentials being used
+2. Not very flexible, as pod will have to startup again if a new secret needs to be loaded
+
+## Handling for the failure cases and logging
+
+- Failure on loading a secret would result in failure on resource creation
+- Failure on authenticating the credentials would result in failure on resource creation
+- Logging should be clear about which credential is being used for a resource
+- Logging should be clear about the credential load/authentication failure
+
+## Milestones
+
+### Support for namespaced secrets
+
+We aim to do the initial work needed in the operator to support this type of multitenancy and add support for namespaced secrets.
+
+**Timelines:** TBD
+
+### Support for per-resource-group secrets
+
+We aim to support per-resource-group secrets under this milestone and configuration work required.
+
+**Timelines:** TBD
+
+### Support for per-resource secrets
+
+We aim to support per-resource secrets under this milestone and configuration work required.
+
+**Timelines:** TBD
+
+### Support for credential cache
+
+TBD
+
+## Open questions
+1. Do we need to support both per-resource and per-resource-group credentials?
+2. Do we want a serviceoperator.azure.com/subscription-id annotation that we write? We can use this to protect against users accidentally updating credential to move resources between subscriptions.
+3. Spell out what fields are in aso-credential. clientid/subscriptionid/tenantId/etc?
+4. How are we going to make this work for Managed Identity?
```

Comment:
I think we should have a meeting with @aramase about this maybe. He might have some good suggestions or raise up some gotchas we haven't thought of.

My initial thinking is something like the following:
1. User creates ManagedIdentity
2. User registers the managed identity Federated identity credential. When they do this, they need to specify the serviceaccount and namespace from which requests will originate. For ASO I think that's always going to be `system:serviceaccount:azureserviceoperator-system:azureserviceoperator-default`... which is a bit awkward because they don't necessarily even have permission to that namespace. We can document it though so it's maybe not the end of the world.
3. User creates a secret in their namespace with their MI details. They may also need to set a flag in that secret like `IsManagedIdentity` or something so that ASO understands this is a managedIdentity and not a SP. Or else we can just use the lack of `ClientSecret` field for that.
4. ASO reads that secret, gets the clientId of the identity it's supposed to use, and uses it somehow. We may need another operator that proxies that identity onto the ASO ServiceAccount, or maybe there's a trick we can use to avoid needing to do that... this is where talking with @aramase would help.

Reply:
Sounds like there's a path forward for this. I will write up the steps and you can incorporate it into this document @super-harsh.

Reply:
Ok so to answer this question, here's what we need to do:

1. ASO must be installed with workload identity. This is because we need access to the Service Account Token, which will only be injected and used if they configure that for the global credential.
2. Users create a `UserManagedIdentity` which they want to use as the identity in a particular namespace. They assign permissions to this identity. This can be done either using ASO itself, or 
3. Users register a `FederatedIdentityCredential` for that `ManagedIdentity`. The `FederatedIdentityCredential` must be registered for the ASO system service account (`"subject": "system:serviceaccount:azureserviceoperator-system:azureserviceoperator-default"`). Note that the users don't need access to our namespace to do this, that subject is basically just a magic string they need to use.
4. User would then create the namespace-scoped `aso-credential` Secret, containing the `clientId` of the User Managed Identity. ASO will read and use that `clientId` when authenticating with Azure.